### ISSUE TITLE
Error when trying to set-relex-server-host from guile shell or cogserver scm shell. (Path to running GHOST)

### ISSUE THREAD

**Issue by Mforrest20** [opencog/opencog#3680]
Hello team, I am attempting to get ghost running in some capacity but am running up against a few problems. The first issue I'm seeing seems to point towards the function: (set-relex-server-host) no longer being recognized.

I am initially importing ghost using: 

`(use-modules (opencog) (opencog nlp) (opencog nlp relex2logic) (opencog openpsi) (opencog ghost) (opencog ghost procedures))`

For the "raw" guile prompt I'm doing and seeeing the following: 

```
scheme@(guile-user)> (use-modules (opencog) (opencog nlp) (opencog nlp relex2logic) (opencog openpsi) (opencog ghost) (opencog ghost procedures))
scheme@(guile-user)> (set-relex-server-host)
;;; <stdin>:11:1: warning: possibly unbound variable `set-relex-server-host'
ice-9/boot-9.scm:1685:16: In procedure raise-exception:
Unbound variable: set-relex-server-host

Entering a new prompt.  Type `,bt' for a backtrace or `,q' to continue.
scheme@(guile-user) [1]> 
```

Doing this with cogserver gives:

```
guile> (use-modules (opencog) (opencog nlp) (opencog nlp relex2logic) (opencog openpsi) (opencog ghost) (opencog ghost procedures))
guile> (set-relex-server-host)
Backtrace:
In ice-9/boot-9.scm:
  1752:10 10 (with-exception-handler _ _ #:unwind? _ # _)
In unknown file:
           9 (apply-smob/0 #<thunk 7fd4d372fc60>)
In ice-9/boot-9.scm:
  1752:10  8 (with-exception-handler _ _ #:unwind? _ # _)
In unknown file:
           7 (apply-smob/0 #<thunk 7fd4d372fc00>)
In ice-9/boot-9.scm:
   2835:4  6 (save-module-excursion #<procedure 7fd4d3731880 at ice-…>)
In ice-9/eval-string.scm:
     38:6  5 (read-and-eval #<input: string 7fd4d2183150> #:lang _)
In ice-9/eval.scm:
   187:27  4 (_ _)
   223:20  3 (proc #<directory (guile-user) 7fd4fc44ac80>)
In unknown file:
           2 (%resolve-variable (7 . set-relex-server-host) #<direct…>)
In ice-9/boot-9.scm:
  1685:16  1 (raise-exception _ #:continuable? _)
In unknown file:
           0 (apply-smob/1 #<exception-handler 7fd4d372fbe0> #<&comp…>)

ERROR: In procedure apply-smob/1:
Unbound variable: set-relex-server-host
ABORT: unbound-variable
```
As I sit here typing this issue I noticed that it appears connected despite giving the above error. For example running the following from a cogserver shell: 

```
guile> (test-ghost "hello")
Error: Cannot connect to RelEx server: unbound-variable
(#f Unbound variable: ~S (LemmaLink) #f)
Backtrace:
In ice-9/boot-9.scm:
  1752:10 10 (with-exception-handler _ _ #:unwind? _ # _)
In unknown file:
           9 (apply-smob/0 #<thunk 7fd4ef0f2500>)
In ice-9/boot-9.scm:
  1752:10  8 (with-exception-handler _ _ #:unwind? _ # _)
In unknown file:
           7 (apply-smob/0 #<thunk 7fd4ef0f2380>)
In ice-9/boot-9.scm:
   2835:4  6 (save-module-excursion #<procedure 7fd4ef0ee6a0 at ice-…>)
In ice-9/eval-string.scm:
     38:6  5 (read-and-eval #<input: string 7fd4d2049a10> #:lang _)
In opencog/ghost/test.scm:
     39:2  4 (test-ghost "hello")
In opencog/ghost.scm:
   318:33  3 (ghost _)
In opencog/nlp/chatbot/chat-utils.scm:
   197:16  2 (nlp-parse _)
In ice-9/boot-9.scm:
  1685:16  1 (raise-exception _ #:continuable? _)
In unknown file:
           0 (apply-smob/1 #<exception-handler 7fd4ef0f22e0> #<&comp…>)

ERROR: In procedure apply-smob/1:
The RelEx server seems to have crashed!
ABORT: misc-error
guile> (test-ghost "goodbye")
Error: Cannot connect to RelEx server: unbound-variable
(#f Unbound variable: ~S (LemmaLink) #f)
Backtrace:
In ice-9/boot-9.scm:
  1752:10 10 (with-exception-handler _ _ #:unwind? _ # _)
In unknown file:
           9 (apply-smob/0 #<thunk 7fd4dd7629c0>)
In ice-9/boot-9.scm:
  1752:10  8 (with-exception-handler _ _ #:unwind? _ # _)
In unknown file:
           7 (apply-smob/0 #<thunk 7fd4dd762960>)
In ice-9/boot-9.scm:
   2835:4  6 (save-module-excursion #<procedure 7fd4dd765c80 at ice-…>)
In ice-9/eval-string.scm:
     38:6  5 (read-and-eval #<input: string 7fd4ef115a80> #:lang _)
In opencog/ghost/test.scm:
     39:2  4 (test-ghost "goodbye")
In opencog/ghost.scm:
   318:33  3 (ghost _)
In opencog/nlp/chatbot/chat-utils.scm:
   197:16  2 (nlp-parse _)
In ice-9/boot-9.scm:
  1685:16  1 (raise-exception _ #:continuable? _)
In unknown file:
           0 (apply-smob/1 #<exception-handler 7fd4dd762940> #<&comp…>)

ERROR: In procedure apply-smob/1:
The RelEx server seems to have crashed!
ABORT: misc-error
```

Results in the relex server displaying:

```
Loop count=5 Restart count=0
Info: Waiting for socket connection
Info: Enter thread with handler 5
Info: hndlr=5 recv input: "hello"
Info: hndlr=5 sentence: "hello"
Link-parsing: 0 milliseconds (avg=0 millisecs, cnt=1)
RelEx processing: 28 milliseconds (avg=28 millisecs, cnt=1)
Info: hndlr=5 sent parse 1 of 1
Info: hndlr=5 Closed input socket
Loop count=6 Restart count=0
Info: Waiting for socket connection
Info: Enter thread with handler 6
Info: hndlr=6 recv input: "goodbye"
Info: hndlr=6 sentence: "goodbye"
Link-parsing: 4 milliseconds (avg=4 millisecs, cnt=1)
RelEx processing: 46 milliseconds (avg=46 millisecs, cnt=1)
Info: hndlr=6 sent parse 1 of 1
Info: hndlr=6 Closed input socket
```
So there does seem to be some connection despite it not appearing to be set.

**Finally** (and this probably belongs in its own issue), but my initial attempts to import the modules in "raw" guile and cogserver resulted in the following errors:

```
scheme@(guile-user)> (use-modules    
 (opencog)
 (opencog nlp)
 (opencog nlp relex2logic)
 (opencog ghost)
 (opencog ghost procedures)
 (opencog exec))
While compiling expression:
In procedure primitive-load-path: Unable to find file "opencog/nlp/relex2logic/loader/load-rules.scm" in load path
```
 and 

```
guile> (use-modules (ice-9 readline)) (activate-readline)
(use-modules (opencog)
             (opencog nlp)
             (opencog nlp relex2logic)
             (opencog openpsi)
             (opencog ghost)
...              (opencog ghost procedures))
Backtrace:
In ice-9/boot-9.scm:
   222:29 19 (map1 _)
   222:17 18 (map1 (((opencog nlp relex2logic)) ((opencog #)) ((…)) …))
  3928:31 17 (_ ((opencog nlp relex2logic)))
  3326:17 16 (resolve-interface (opencog nlp relex2logic) #:select _ …)
In ice-9/threads.scm:
    390:8 15 (_ _)
In ice-9/boot-9.scm:
  3252:13 14 (_)
In ice-9/threads.scm:
    390:8 13 (_ _)
In ice-9/boot-9.scm:
  3536:20 12 (_)
   2835:4 11 (save-module-excursion _)
  3556:26 10 (_)
In unknown file:
           9 (primitive-load-path "opencog/nlp/relex2logic" #<proced…>)
In opencog/nlp/relex2logic.scm:
    110:0  8 (_)
    104:8  7 (load-r2l-rulebase)
In ice-9/boot-9.scm:
   2835:4  6 (save-module-excursion #<procedure 7fd4ec16a450 at ice-…>)
   4394:9  5 (_)
   2188:5  4 (%start-stack load-stack _)
   2190:9  3 (_)
In unknown file:
           2 (primitive-load-path "opencog/nlp/relex2logic/loader/lo…")
In ice-9/boot-9.scm:
  1685:16  1 (raise-exception _ #:continuable? _)
In unknown file:
           0 (apply-smob/1 #<exception-handler 7fd4eee1fa00> #<&comp…>)

ERROR: In procedure apply-smob/1:
In procedure primitive-load-path: Unable to find file "opencog/nlp/relex2logic/loader/load-rules.scm" in load path
ABORT: misc-error
```
However they are loaded on subsequent load module requests. I'm not sure if this is an artifact or running it a second time is obscuring some other problem.

Thank you @linas and @AmeBel for any help you can provide with these issues, and context surrounding the odd behavior from imports.

**Comment by linas** [comment]
What a snowball. Some remarks:
* The `(set-relex-server-host)` is defined in module `(opencog oc)` so you have to `(use-modules (opencog oc))` to get it.
* During debugging it will be easier to load the modules one at a time.  There is no reason to load them all at once, except less typing.
* ``LemmaLink` is defined in `oc` above. 
* `The RelEx server seems to have crashed!` -- I assume it hasn't actually crashed, and that its still running and usable!? I assume that the missing `LemmaLink` is reported as a crashed server.
*  import the modules in "raw" guile -- that should have worked. Let me check.

**Comment by linas** [comment]
I miss-typed, it should be `(use-modules (opencog nlp oc)) `

**Comment by linas** [comment]
> import the modules in "raw" guile

Works for me. What version of guile are you using? `guile --version` -- it should be version 3.01 or better.  Everything should have been compiled with `guile-3.0-dev` installed -- if you have `guile-2.2-dev` you'll want to remove that. `apt list --installed |grep guile` for debian-based systems; I don't know the equivalent for RedHat.  (unless you installed guile from source!?)

**Comment by Mforrest20** [comment]
I'm about to try your recommendation. 

To give you a quick update to your questions:

1. Your assumption is correct, it did **not** crash. It seems to close the socket after handling each request. (presumably the expected behavior)
2. With regards to my guile version, it is guile (GNU Guile) 3.0.7. Running Ubuntu 20.04 in a VM, and installed guile with apt.

**Comment by linas** [comment]
> In procedure primitive-load-path: Unable to find file "opencog/nlp/relex2logic/loader/load-rules.scm" in load path

I can now reproduce this ...I think I see the bug (s)...

**Comment by linas** [comment]
> In procedure primitive-load-path: Unable to find file "opencog/nlp/relex2logic/loader/load-rules.scm" in load path

I think I just fixed this, in commit 008cdbe7f192fc66b3deadf6400f8edb54533f36

relex2logic was making poor assumptions about modules: the name of the module must be the same as the name of the file that defines the module. Otherwise, things don't work. So that's patched up in that commit.  Pull and reinstall and it should work.  Just pull this repo, all the other repos are OK.

**Comment by linas** [comment]
I'm going to close this bug, because I think the fix of 008cdbe7f192fc66b3deadf6400f8edb54533f36 will have fixed everything reported here.  Open a new issue if problems persist,.
